pyteal/ast/abi/uint.py

```diff
@@ -0,0 +1,99 @@
+from typing import Union, cast
+from abc import abstractmethod
+
+from ...types import TealType
+from ..expr import Expr
+from ..seq import Seq
+from ..assert_ import Assert
+from ..substring import Suffix
+from ..int import Int
+from ..unaryexpr import Itob, Not
+from ..binaryexpr import ExtractUint64, ExtractUint16
+from .type import Type
+
+NUM_BITS_IN_BYTE = 8
+
+
+class Uint(Type):
+    def __init__(self, bit_size: int) -> None:
+        valueType = TealType.uint64 if bit_size <= 64 else TealType.bytes
+        super().__init__(valueType)
+        self.bit_size = bit_size
+
+    def has_same_type_as(self, other: Type) -> bool:
+        return isinstance(other, Uint) and self.bit_size == other.bit_size
+
+    def is_dynamic(self) -> bool:
+        return False
+
+    def bits(self) -> int:
+        return self.bit_size
+
+    def byte_length_static(self) -> int:
+        return self.bit_size // NUM_BITS_IN_BYTE
+
+    def get(self) -> Expr:
+        return self.stored_value.load()
+
+    @abstractmethod
+    def set(self, value: Union[int, Expr]) -> Expr:
+        pass
+
+
+class Uint16(Uint):
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__(16)
+
+    def new_instance(self) -> "Uint16":
+        return Uint16()
+
+    def set(self, value: Union[int, Expr]) -> Expr:
+        if type(value) is int:
+            if value >= 2 ** 16:
+                raise ValueError("Value exceeds Uint16 maximum: {}".format(value))
+            value = Int(value)
+        # TODO: check dynamic value bounds?
+        return self.stored_value.store(cast(Expr, value))
+
+    def decode(self, encoded: Expr, offset: Expr, length: Expr) -> Expr:
+        return Seq(
+            Assert(length == Int(self.byte_length_static())),  # TODO: remove?
+            self.set(ExtractUint16(encoded, offset)),
+        )
+
+    def encode(self) -> Expr:
+        # value might exceed a uint16, need to check at runtime
+        return Seq(
+            Assert(Not(self.get() >> Int(2))),
```

Comment:
```suggestion
            Assert(Not(self.get() >> Int(16))),
```

Reply:
Good catch! Though why 17 and not 16? `2**16 >> 16 = 1`

Reply:
Yes exactly. Furthermore, if every bit until the 16th is set, `2**17-1 >> 16 = 1`

Reply:
I think we disagree by 1. The maximum value of a 16-bit unsigned int is `2**16-1 = 65535`. So it should be sufficient to right shift by 16 to see if anything is equal to or greater than that.

My original code meant 2 bytes, but of course shifting is measured in bits, not bytes, so it should be 16.

Reply:
Ah yes you are right. I didn't have my engineering hat on.